File: 1894-MergeStringsAlternately/1894-MergeStringsAlternately.cpp
// Last updated: 6/4/2025, 9:19:52 PM
#include <string>
using namespace std;

class Solution {
public:
    string mergeAlternately(string word1, string word2) {
        string merged;
        merged.reserve(word1.length() + word2.length()); // Reserve space for the merged string
        
        auto it1 = word1.begin(), it2 = word2.begin();
        const auto end1 = word1.end(), end2 = word2.end();
        
        // Merge the strings while there are characters in both words
        while (it1 != end1 && it2 != end2) {
            merged.push_back(*it1++);
            merged.push_back(*it2++);
        }
        
        // Append the remaining characters from word1, if any
        while (it1 != end1) {
            merged.push_back(*it1++);
        }
        
        // Append the remaining characters from word2, if any
        while (it2 != end2) {
            merged.push_back(*it2++);
        }
        
        return merged;
    }
};

File: 1894-MergeStringsAlternately/1894-MergeStringsAlternately.py
# Last updated: 6/4/2025, 9:19:36 PM
class Solution:
    def mergeAlternately(self, word1, word2):
        # Initialize an empty list to store the characters of the merged string
        merged = []
        
        # Iterate through both strings simultaneously
        i = 0
        while i < len(word1) and i < len(word2):
            merged.append(word1[i])
            merged.append(word2[i])
            i += 1
        
        # Append any remaining characters from word1
        if i < len(word1):
            merged.extend(word1[i:])
        
        # Append any remaining characters from word2
        if i < len(word2):
            merged.extend(word2[i:])
        
        # Convert the list to a string and return it
        return ''.join(merged)

# Test cases
solution = Solution()
print(solution.mergeAlternately("abc", "pqr"))   # Output: "apbqcr"
print(solution.mergeAlternately("ab", "pqrs"))   # Output: "apbqrs"
print(solution.mergeAlternately("abcd", "pq"))   # Output: "apbqcd"
